### ISSUE TITLE
Query Object reference not set to an instance of an object

### ISSUE THREAD

**Issue by brunoalreis** [aspnet/EntityFrameworkCore#17091]
I am making a reusable query for difference scenarios.
When i do a N to M query, if the M has other optional relation, it gives me an Object reference not set to an instance of an object exception.
This only happens, if i split the query in two parts. See the GetPersonsNotWorking2 bellow.

[Bug.zip](https://github.com/aspnet/EntityFrameworkCore/files/3492147/Bug.zip)

I've created a simple example of the problem: I have this tables:
 - Group - Represents a group of persons
 - PersonGroup - Relation between person and group (N to M relation)
 - Person - Represents a persons
 - Job - Represents a person job. A person may not have a job.

```
Exception message:
Stack trace:
Object reference not set to an instance of an object.
   at System.Linq.Enumerable.SelectEnumerableIterator`2.MoveNext()
   at Microsoft.EntityFrameworkCore.Query.Internal.LinqOperatorProvider.ExceptionInterceptor`1.EnumeratorExceptionInterceptor.MoveNext()
   at System.Collections.Generic.List`1.AddEnumerable(IEnumerable`1 enumerable)
   at System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)
   at Bug.DbManager.GetPersonsNotWorking2() in D:\Projectos\Bug\DbManager.cs:line 103
   at Bug.Program.Main(String[] args) in D:\Projectos\Bug\Program.cs:line 38
```

### Steps to reproduce
```c#
    class Group
    {
        [Key]
        public int Id { get; set; }
        public string Name { get; set; }
    }

    class PersonGroup
    {
        [Key]
        public int Id { get; set; }
        public int IdPerson { get; set; }
        public int IdGroup { get; set; }

        #region Navigation Properties
        public Group Group { get; set; }
        public Person Person { get; set; }
        #endregion
    }

    class Person
    {
        [Key]
        public int Id { get; set; }
        public string Name { get; set; }
        public int? IdJob { get; set; }


        #region Navigation Properties
        public Job Job { get; set; }
        #endregion
    }

    class Job
    {
        [Key]
        public int Id { get; set; }
        public string Name { get; set; }
    }

        public List<PersonIds> GetPersonsWorking()
        {
            return
                GetFullPersonQueryWorking()
                    .Select(x => new PersonIds
                    {
                        IdPerson = x.Person.Id,
                        IdJob = x.Job.Id,
                        IdGroup = x.Group.Id
                    })
                    .ToList();
        }

        private IQueryable<FullPerson> GetFullPersonQueryWorking()
        {
            return
                from dbgroup in _context.Groups

                join personGroup in _context.PersonGroup
                    on dbgroup.Id equals personGroup.IdGroup

                join person in _context.Persons
                    on personGroup.IdPerson equals person.Id

                join job in _context.Jobs
                    on person.IdJob equals job.Id
                    into Jobs
                from job in Jobs.DefaultIfEmpty()

                select new FullPerson
                {
                    Group = dbgroup,
                    Person = person,
                    Job = job,
                };
        }

        public List<PersonIds> GetPersonsNotWorking2()
        {
            var getPersonAndJobQuery =
                from person in _context.Persons

                join job in _context.Jobs
                    on person.IdJob equals job.Id
                    into Jobs
                from job in Jobs.DefaultIfEmpty()

                select new
                {
                    Person = person,
                    Job = job,
                };

            var getFullPersonQuery =
                from dbgroup in _context.Groups
                join personGroup in _context.PersonGroup
                    on dbgroup.Id equals personGroup.IdGroup
                join personAndJob in getPersonAndJobQuery
                    on personGroup.IdPerson equals personAndJob.Person.Id
                select new
                {
                    Person = personAndJob.Person,
                    Job = personAndJob.Job,
                    Group = dbgroup
                };

            return
                getFullPersonQuery 
                    .Select(x => new PersonIds
                    {
                        IdPerson = x.Person.Id,
                        IdJob = x.Job.Id,
                        IdGroup = x.Group.Id
                    })
                    .ToList();
        }

```

### Further technical details
EF Core version: 2.2.4 (in my real project is 2.2.6)
Database Provider: Microsoft.EntityFrameworkCore.PostgreSQL
Operating system: Windows 10
IDE: Visual Studio 2019 16.2

**Comment by ajcvickers** [comment]
This may work in 3.0. Moving to the backlog to investigate.

**Comment by brunoalreis** [comment]
@ajcvickers, thank you for the answer. I am running ef core on .net framework and in ef core 3.0 will not be compatible with it no more. Any chance of this will be released on ef core 2 update?

**Comment by ajcvickers** [comment]
@brunoalreis We are not currently planning to patch this in any 2.x release.

**Comment by ajcvickers** [comment]
Confirmed this issue is fixed in EF Core 3.1.
